src/main/java/com/google/api/codegen/ApiaryConfig.java

```diff
@@ -102,15 +102,59 @@
   private final Table<Type, String, Field> fields = HashBasedTable.<Type, String, Field>create();
 
   /*
-   * The service canonical name, or name if no canonical name
+   * The service canonical name, or name if no canonical name.
    */
   private String serviceCanonicalName;
 
   /*
-   * The service version string
+   * The service version string.
    */
   private String serviceVersion;
 
+  private Map<String, AuthType> authOverrides = new HashMap<>();
+
+  /*
+   * If present in the scope list, indicates that the API supports application default credentials
+   * based auth.
+   */
+  private static final String CLOUD_PLATFORM_SCOPE =
+      "https://www.googleapis.com/auth/cloud-platform";
+
+  /*
+   * Possible auth types supported by discovery.
+   */
+  public enum AuthType {
+    APPLICATION_DEFAULT_CREDENTIALS,
+    OAUTH_3L,
+    API_KEY
+  }
+
+  /*
+   * Returns the auth type supported by the service.
+   */
+  public AuthType getAuthType() {
+    String key = getServiceCanonicalName();
+    if (authOverrides.containsKey(key)) {
+      return authOverrides.get(key);
+    }
+    // This statement is based on the assumption that every method in a service contains all the
```

Comment:
I wonder whether this assumption can land us in trouble.

I also wonder whether it can be misleading. What if the first method is a write method, and so it doesn't list any `readonly` scopes? Then when we generate for a read method, we don't list the `readonly` scopes, even though it would work. Thus, the user thinks that any of the listed scopes work but doesn't realize that the `readonly` scope would, too.

Also, what if the first method has a set of resource-specific scopes, but there are other resources in the API? eg: `mail.thread.readonly` vs `mail.contact.readonly`, say. (I made that up, btw.)

I think it's safest to do this per method....